src/fs_storage.c

```diff
@@ -0,0 +1,77 @@
+/*
+ *   Software Updater - client side
+ *
+ *      Copyright © 2012-2018 Intel Corporation.
+ *
+ *   This program is free software: you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation, version 2 or later of the License.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <sys/statvfs.h>
+
+#include "swupd.h"
+
+/* recursively calculate size from a complete list of bundle_result structs */
+long calculate_size(char *bname, struct list *bundle_info, bool installed)
+{
+	struct list *ptr;
+	struct bundle_result *bi;
+	long size = 0;
+	ptr = list_head(bundle_info);
+	while (ptr) {
+		bi = ptr->data;
+		ptr = ptr->next;
+		if (bi->seen || strcmp(bname, bi->bundle_name) != 0) {
+			continue;
+		}
+
+		/* If installed is true the initial call to this recursive
+		 * function was for an installed bundle. In this case we want
+		 * to calculate the contentsize of all included bundles (we
+		 * know they are all installed already) so we can report the
+		 * total installed size on the system.
+		 *
+		 * Otherwise only add the contentsize of bundles not already
+		 * installed on the system. */
+		if (installed || !is_tracked_bundle(bname)) {
+			size += bi->topsize;
+			bi->seen = true;
+		}
+
+		/* add bundle sizes for includes as well */
+		struct list *ptr2;
+		ptr2 = list_head(bi->includes);
+		while (ptr2) {
+			char *inc = NULL;
+			inc = ptr2->data;
+			ptr2 = ptr2->next;
+			/* recursively add included sizes */
+			size += calculate_size(inc, bundle_info, installed);
```

Comment:
There is one possible infinite recursion loop if you have an installed bundle that includes itself. I know that this won't happen in a system, but anyway this is a problem in the API of this function. You can fix this by moving the "bi->seen = true;" out of the if. I think this won't affect any other usages of this function, but it's better to check that. And It's pretty easy to get rid of this recursion too because it doesn't need to go in more than one level.

Reply:
Include loops are strictly prohibited in bundle generation (in various validation checks and in mixer code itself) as they cause many more issues than the issue you point out here. That said, I have no issue with fixing it here.